### PR TITLE
[Ready for Review] Universal Object Damage System Phase 1: Implementation

### PR DESCRIPTION
**Pull request by Couls**
**What does this PR do:**
Ports over TG's universal damage system which can be found here: https://github.com/tgstation/tgstation/pull/20793
This new system adds the new acid damage type for objects and gives them all a system where they can be destroyed and damaged. I also added in beakers and drinks being splashed on people when thrown at and shattering.

sneak peek of the destruction that can be caused by a fluorosulphuric acid cluster bang grenade:
image

I plan to do this in multiple phases, The first phase being implementation and then the second phase being balance. It will be easier to balance and configure this after we see how its handled live.

New fire sprite:
687474703a2f2f69636f6e64696666626f742e64657874726173706163652e6e65743a383038302f46696c65732f32373238363435322f3132393731363932302f31362f6265666f72652e676966

The system is now implemented, balance is something that will need to be worked on for sure.

**Changelog:**
:cl:
add: Object damage overhaul, new acid damage, mechs, blobs, tspiders and xeno structures are all handled differently now.
add: Clothes can now be damaged and burned
add: Beakers now shatter and spill their contents over whatever they're thrown on.
add: mechs now take increased damage from behind and decreased damage from infront
/:cl:

### CODE REVIEW

**Comment by taukausanake** [comment]
> Beakers now shatter and spill their contents over whatever they're thrown on

Ghetto chem grenades. I like.
The beakers won't spill if sealed, yes? They can at least shatter still if not caught by someone else. I really like this in that case

**Comment by Ty-Omaha** [comment]
I want objects to be destroyed in fires

**Comment by Markolie** [comment]
Amazing work!

**Comment by Couls** [comment]
> I want objects to be destroyed in fires

most objects can and will be and will turn to ash, caught in acid will turn to goop

> Ghetto chem grenades. I like.
> The beakers won't spill if sealed, yes? They can at least shatter still if not caught by someone else. I really like this in that case

I'll include a check for lid in the toss proc

**Comment by Bookdrizzt** [comment]
Will Borgs retain their fireproofing? Also, YEEEEEEESSSSS!!! Finally, we can have options BESIDES shooting and blowing stuff up to destroy objects!!!

**Comment by Couls** [comment]
Borgs aren't objects so they still treat damage the same, this goes for all mobs everyone still gets hurt all the same the only difference is objects and structures(which actually includes mechs)

File: code/controllers/subsystem/acid.dm [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	name = "Acid"
+	priority = FIRE_PRIORITY_ACID
+	flags = SS_NO_INIT|SS_BACKGROUND
+	runlevels = RUNLEVEL_GAME | RUNLEVEL_POSTGAME
+
+	var/list/currentrun = list()
+	var/list/processing = list()
+
+/datum/controller/subsystem/acid/stat_entry()
+	..("P:[processing.len]")
```
Comment (Fox-McCloud) [inline code/controllers/subsystem/acid.dm]: Make this something like "melting objects" or something more clear.

File: code/game/atoms.dm [hunk 2]
```diff
@@ -160,10 +160,9 @@
 
 /atom/proc/attack_hulk(mob/living/carbon/human/user, does_attack_animation = FALSE)
```
Comment (Fox-McCloud) [inline code/game/atoms.dm]: is `does_attack_animation` used anywhere anymore, then?

File: code/game/atoms.dm [hunk 3]
```diff
@@ -764,3 +766,8 @@ var/list/blood_splatter_icons = list()
 
 /atom/Exited(atom/movable/AM, atom/newLoc)
 	SEND_SIGNAL(src, COMSIG_ATOM_EXITED, AM, newLoc)
+
+/atom/proc/mech_melee_attack(obj/mecha/M)
+	return
```
Comment (Fox-McCloud) [inline code/game/atoms.dm]: new line.

File: code/game/atoms.dm [hunk 4]
```diff
@@ -764,3 +766,8 @@ var/list/blood_splatter_icons = list()
 
 /atom/Exited(atom/movable/AM, atom/newLoc)
 	SEND_SIGNAL(src, COMSIG_ATOM_EXITED, AM, newLoc)
+
+/atom/proc/mech_melee_attack(obj/mecha/M)
+	return
+/atom/proc/contents_explosion(severity, target)
```
Comment (Fox-McCloud) [inline code/game/atoms.dm]: our explosions system doesn't work with this, at all.

File: code/game/gamemodes/blob/theblob.dm [hunk 5]
```diff
@@ -18,11 +20,18 @@
 	blobs += src
 	src.dir = pick(1, 2, 4, 8)
 	src.update_icon()
-	..(loc)
+	..()
+	ConsumeTile()
+	return
```
Comment (Fox-McCloud) [inline code/game/gamemodes/blob/theblob.dm]: trailing return

File: code/game/machinery/camera/camera.dm [hunk 6]
```diff
@@ -233,10 +244,9 @@
 	. = ..()
 
 /obj/machinery/camera/obj_break(damage_flag)
-	if(status)
+	if(status && !(flags & NODECONSTRUCT))
 		triggerCameraAlarm()
-		toggle_cam(null, FALSE)
-		wires.CutAll()
```
Comment (Fox-McCloud) [inline code/game/machinery/camera/camera.dm]: TG doesn't have wires for cameras (which IMO is a good thing), but we do, so this needs to stay.

File: code/game/machinery/computer/ai_core.dm [hunk 7]
```diff
@@ -4,15 +4,20 @@
...
 /obj/structure/AIcore/Destroy()
-	QDEL_NULL(laws)
-	QDEL_NULL(circuit)
-	QDEL_NULL(brain)
+	if(circuit)
+		qdel(circuit)
+		circuit = null
+	if(brain)
+		brain.forceMove(loc)
+		brain = null
```
Comment (Fox-McCloud) [inline code/game/machinery/computer/ai_core.dm]: what existed before was superior. The brain change is a balance change though.

File: code/game/machinery/computer/honkputer.dm [hunk 8]
```diff
@@ -10,7 +10,7 @@
 	circuit = /obj/item/circuitboard/HONKputer
 	var/authenticated = 0
 	var/message_cooldown = 0
-	var/state = STATE_DEFAULT
+	//var/state = STATE_DEFAULT
```
Comment (Fox-McCloud) [inline code/game/machinery/computer/honkputer.dm]: no commented out code except for exceptional reason.

File: code/game/machinery/constructable_frame.dm [hunk 9]
```diff
@@ -218,6 +218,18 @@
 				if(!success)
 					to_chat(user, "<span class='danger'>You cannot add that to the machine!</span>")
 					return 0
+	if(user.a_intent == "harm")
+		return ..()
+
+
+/obj/machinery/constructable_frame/machine_frame/deconstruct(disassembled = TRUE)
```
Comment (Fox-McCloud) [inline code/game/machinery/constructable_frame.dm]: we don't have these anywhere.

File: code/game/machinery/deployable.dm [hunk 10]
```diff
@@ -58,63 +58,30 @@ for reference:
...
-/obj/structure/barricade/blob_act()
-	src.health -= 25
-	if(src.health <= 0)
-		visible_message("<span class='danger'>The blob eats through \the [src]!</span>")
-		qdel(src)
+/obj/structure/barricade/proc/make_debris()
 	return
 
+/obj/structure/barricade/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/weldingtool) && user.a_intent != "harm")
```
Comment (Fox-McCloud) [inline code/game/machinery/deployable.dm]: use helpers

File: code/modules/reagents/chemistry/reagents/toxins.dm [hunk 11]
```diff
@@ -240,56 +240,38 @@
...
 	taste_message = "<span class='userdanger'>ACID</span>"
 
-/datum/reagent/sacid/on_mob_life(mob/living/M)
+/datum/reagent/acid/on_mob_life(mob/living/M)
 	var/update_flags = STATUS_UPDATE_NONE
 	update_flags |= M.adjustFireLoss(1, FALSE)
 	return ..() | update_flags
 
-/datum/reagent/sacid/reaction_mob(mob/living/M, method = TOUCH, volume)
+/datum/reagent/acid/reaction_mob(mob/living/M, method = TOUCH, volume)
```
Comment (Fox-McCloud) [inline code/modules/reagents/chemistry/reagents/toxins.dm]: leave both acid's `reaction_mob` alone. This is Goonchem, not TG chem.

File: code/modules/reagents/chemistry/reagents/toxins.dm [hunk 12]
```diff
@@ -240,56 +240,38 @@
...
-				H.emote("scream")
-
-/datum/reagent/sacid/reaction_obj(obj/O, volume)
-	if((istype(O,/obj/item) || istype(O,/obj/structure/glowshroom)) && prob(40))
-		if(!O.unacidable)
-			var/obj/effect/decal/cleanable/molten_object/I = new/obj/effect/decal/cleanable/molten_object(O.loc)
-			I.desc = "Looks like this was \an [O] some time ago."
-			O.visible_message("<span class='warning'>[O] melts.</span>")
-			qdel(O)
+/datum/reagent/acid/reaction_obj(obj/O, volume)
```
Comment (Fox-McCloud) [inline code/modules/reagents/chemistry/reagents/toxins.dm]: This is fine though.

File: code/modules/mob/living/carbon/update_icons.dm [hunk 13]
```diff
@@ -34,6 +34,8 @@
 /mob/living/carbon/proc/handle_transform_change()
 	return
 
+/obj/item/proc/worn_overlays(isinhands = FALSE, icon_file)
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/update_icons.dm]: not sure this even works.

**Comment by KasparoVy** [comment]
Aha, now we can destroy the nuke! It was as the prophets of goone foretold.

**Comment by Fox-McCloud** [comment]
This is a lot of blind copy-pasta.

While the work is *very* much appreciated, you need to analyze the system on both our end and TG's end to have an understanding of what it does/how, in a number of places, because just copy-pasting over a number of these things is going to break things, horrifically.

**Comment by Couls** [comment]
Thanks for the review fox, I'm still actively working on it and testing it I just wanted to have it up while it was working(read: compiling) so I could get stuff like that review, feedback, information on how sturdy and such we want things to be for situations where I can't really refer to TG like with terror spiders and that such. I will continue working on it and fixing any obvious issues that come up

**Comment by KasparoVy** [comment]
So long as I can catch beakers of BS by having throw intent on 24/7.

File: code/game/atoms.dm (hunk 2 above)
Comment (Couls) [reply code/game/atoms.dm]: No it doesn't seem like it, I'll remove it from the arguments

File: code/modules/mob/living/carbon/update_icons.dm (hunk 13 above)
Comment (Couls) [reply code/modules/mob/living/carbon/update_icons.dm]: works when I tested it

**Comment by taukausanake** [comment]
I'd rather there be a timeout period after setting throw intent when you can catch anything but that's for another PR

**Comment by Quantum-M** [comment]
This means acid can destroy objects, like boxes and such, right?

If so, one should probably look into making weapon secure cases you can make from RnD resist the acid, because it will just become a ghetto way to get weapons on part of science (even as a non-antag).

**Comment by taukausanake** [comment]
Oh oh! Titanium cases! I'm not too sure how resistant titanium is to acids (a quick Google search says maybe a lot) but just make all secure items printable from an auto or protolathe require titanium, maybe one or two sheets, just to make the case and then you can deconstruct it back into the titanium it was made out of. The cases can then be made resistant in the code and claimed resistant by virtue of their mineral.

I really hope this is a good idea, I used so many brain cells to make it

**Comment by Couls** [comment]
I’ll  test how long it takes to open a lockbox with acid and then see about making them acidproof, I feel like if cargo can use emitters to break open security gun crates science should be able to make acid it’s pretty loud so people in the area can probably hear when it’s being used and using too much risks breaking the box and destroying everything inside

File: code/game/machinery/constructable_frame.dm (hunk 9 above)
Comment (Couls) [reply code/game/machinery/constructable_frame.dm]: fixed

File: code/game/machinery/computer/ai_core.dm (hunk 7 above)
Comment (Couls) [reply code/game/machinery/computer/ai_core.dm]: reverted it

**Comment by Citinited** [comment]
Does this PR refactor the deflection chance that mecha and (I think) cyborgs have? If not then I think a refactor of those parts of the code would fit right into this PR

**Comment by Couls** [comment]
I left most of the values for deflection the same but now mech damage handling is all moved to its own file and easily modifiable if someone wants to overhaul how its handled. Mechs now take double damage from behind and 1.5x less damage from infront due to their plating, to be completely honest I only know the basics of how the mech code works so I can't explain it in detail myself

**Comment by Couls** [comment]
Let me know if you guys want to keep or remove the whole thing with beakers and glass smashing when thrown

**Comment by TDSSS** [comment]
New meta ultimate throwing weapon: Large beakers of acid.

**Comment by taukausanake** [comment]
Just make sure you hit your target

**Comment by Couls** [comment]
The new hard counter to that would be carrying a beaker of water

**Comment by Shadeykins** [comment]
> 
> 
> I’ll test how long it takes to open a lockbox with acid and then see about making them acidproof, I feel like if cargo can use emitters to break open security gun crates science should be able to make acid it’s pretty loud so people in the area can probably hear when it’s being used and using too much risks breaking the box and destroying everything inside.
> 
> As an update to the status of this PR, I have a few more objects to go through and then I’ll start working on ensuring everything is paradise friendly starting with terror spiders and then will work through some bugs in testing before it’ll be ready for review

Lockboxes are deliberately designed to prevent science players from accessing those items without an emag. Their inclusion is a deliberate design choice to prevent illegitimate non-traitor access to these things without the HoS, Warden, or Captain being there whatsoever.

**Comment by Couls** [comment]
Hmm okay i'll make the science ones indestructible but still make it possible to break open captains lockbox

**Comment by Bookdrizzt** [comment]
Actually, is clothing damage part of the stuff being ported at this stage? Because if so, that makes me happy that being set on fire has literally 0 consequences if you have a silver sulfazine or synthflesh. Plus, as sec, if someone sets themselves on fire, now I could arrest them for indecent exposure, too!

**Comment by Couls** [comment]
First stage implements most basic systems, the next phase is mainly balancing and dealing with special cases

**Comment by SkeletalElite** [comment]
My favorite part about universal object damage is breaking the machine in perma in order to make a spear, who needs to risk shocking yourself.

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/terror_spiders.dm [hunk 14]
```diff
@@ -198,7 +198,7 @@ var/global/list/ts_spiderling_list = list()
 				F.open()
 		else
 			to_chat(src, "Closing fire doors does not help.")
-	else if(istype(target, /obj/machinery/door/airlock))
+	else if(istype(target, /obj/machinery/door/airlock) && a_intent == INTENT_HELP)
```
[the remark this answers is not in the document]
Comment (Couls) [reply code/modules/mob/living/simple_animal/hostile/terror_spiders/terror_spiders.dm]: Yep!

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/black.dm [hunk 15]
```diff
@@ -21,6 +21,7 @@
 	health = 120
 	melee_damage_lower = 5
 	melee_damage_upper = 10
+	obj_damage = 100
```
[the remark this answers is not in the document]
Comment (Couls) [reply code/modules/mob/living/simple_animal/hostile/terror_spiders/black.dm]: Yeah exactly, its seperate from melee damage

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/green.dm [hunk 16]
```diff
@@ -20,6 +20,7 @@
 	health = 150
 	melee_damage_lower = 10
 	melee_damage_upper = 20
+	obj_damage = 200
```
[the remark this answers is not in the document]
Comment (Couls) [reply code/modules/mob/living/simple_animal/hostile/terror_spiders/green.dm]: yeah thats an oversight on my part, will fix that right now

**Comment by Couls** [comment]
Let’s bring this back boy back

**Comment by Couls** [comment]
I'm just gonna redo this whole thing

**Comment by Ty-Omaha** [comment]
oof

**Comment by KasparoVy** [comment]
The toughest choices require the strongest wills bro.

**Pull request closed without merging**